### PR TITLE
[FIX] payment_authorize: Deprecate redirect flow

### PR DESCRIPTION
**Pull request by jvh83**
After Authorize documentation, the redirection flow
as developed in Odoo is not supported anymore.
See https://developer.authorize.net/api/upgrade_guide.html#HTTP_GET

This is a try without touching any views to avoid
the need to update the module.

OPW-2671238

### CODE REVIEW

File: addons/payment_authorize/models/payment.py [hunk 1]
```diff
@@ -188,6 +194,19 @@ def authorize_test_credentials(self):
         transaction = AuthorizeAPI(self.acquirer_id)
         return transaction.test_authenticate()
 
+    @api.multi
+    def write(self, values):
+        super(PaymentAcquirerAuthorize, self).write(values)
+        self._check_authorize_payment_flow()
```
Comment (Feyensv) [inline addons/payment_authorize/models/payment.py:200]:
```suggestion
        res = super(PaymentAcquirerAuthorize, self).write(values)
        # force the constraint check for any update on acquirers to alert
        # as soon as possible if the acquirer cannot work in its current state
        self._check_authorize_payment_flow()
        return res
```

* write calls return a boolean.  It doesn't mean anything atm, but API should be consistent anyway.
* we can avoid the manual call to the constraint when it was already called by the orm.

File: addons/payment_authorize/models/payment.py [hunk 2]
```diff
@@ -27,6 +27,12 @@ class PaymentAcquirerAuthorize(models.Model):
     authorize_transaction_key = fields.Char(string='API Transaction Key', required_if_provider='authorize', groups='base.group_user')
     authorize_signature_key = fields.Char(string='API Signature Key', groups='base.group_user', compute="_compute_auth_signature_key", inverse="_inverse_auth_signature_key")
 
+    @api.constrains('payment_flow')
+    def _check_authorize_payment_flow(self):
+        for acquirer in self.filtered(lambda x: x.provider == "authorize"):
+            if acquirer.payment_flow == 'form':
+                raise UserError(_('The "Redirection to the acquirer website" payment flow is deprecated for Authorize.net. Please choose the "Payment From Odoo" flow.'))
```
Comment (Feyensv) [inline addons/payment_authorize/models/payment.py:34]: Maybe avoid raising an error for inactive acquirers?

File: addons/payment_authorize/models/payment.py (hunk 1 above)
Comment (jev-odoo) [reply addons/payment_authorize/models/payment.py]: Indeed, much better. Thank you for your insight

File: addons/payment_authorize/models/payment.py (hunk 2 above)
Comment (jev-odoo) [reply addons/payment_authorize/models/payment.py:34]: Last fix should be enough to filter active acquirer

File: addons/payment_authorize/models/payment.py [hunk 3]
```diff
@@ -29,9 +29,12 @@ class PaymentAcquirerAuthorize(models.Model):
 
     @api.constrains('payment_flow')
     def _check_authorize_payment_flow(self):
-        for acquirer in self.filtered(lambda x: x.provider == "authorize" and x.website_published):
-            if acquirer.payment_flow == 'form':
-                raise UserError(_('The "Redirection to the acquirer website" payment flow is deprecated for Authorize.net. Please choose the "Payment From Odoo" flow.'))
+        form_authorize_acquirers = self.filtered(lambda x: x.provider == "authorize" and x.payment_flow == "form")
+        if form_authorize_acquirers:
+            form_string = [t[1] for t in self._fields['payment_flow']._description_selection(self.env) if t[0] == 'form'][0]
+            s2s_string = [t[1] for t in self._fields['payment_flow']._description_selection(self.env) if t[0] == 's2s'][0]
```
Comment (Feyensv) [inline addons/payment_authorize/models/payment.py:35]:
```suggestion
            payment_flow_desc = {value: desc for value,desc in self._fields['payment_flow']._description_selection(self.env)
            form_string = payment_flow_desc['form']
            s2s_string = payment_flow_desc['s2s']
```

?

Comment (jev-odoo) [reply addons/payment_authorize/models/payment.py]: Done ;)

**Comment by jev-odoo** [comment]
For info, the redirect flow has been removed for Authorize as from saas-14.3

**Comment by jorenvo** [comment]
Why can't we follow the deprecation advice and change the verb to POST? It's too complex?

**Comment by AntoineVDV** [comment]
> Why can't we follow the deprecation advice and change the verb to POST? It's too complex?

@jorenvo AFAIK the redirect requests in 12.0 are already POST. Any clue why POST requests also fail @jev-odoo?

**Pull request closed without merging**